File: android/app/src/main/java/it/curzel/sneakbit/NativeState.kt
package it.curzel.sneakbit

import android.content.Context
import android.util.Log
import org.json.JSONObject
import java.io.File

/**
 * The Android half of the native save bridge (see js/nativeBridge.js) — the
 * counterpart of electron/nativeState.js and ios/SneakBit/NativeState.swift.
 *
 * Two files in internal storage matter:
 *
 *   filesDir/save.json           the save left by the Rust build this one
 *                                replaces. This ships as an update to the same
 *                                applicationId (it.curzel.bitscape), so that
 *                                directory survives the update and the web
 *                                build can import it once. Never deleted — it's
 *                                the player's only copy of a save that predates
 *                                accounts and cloud sync.
 *
 *   filesDir/sneakbit-save.json  this build's own save, mirrored out of
 *                                localStorage. WebView's storage is opaque and
 *                                has no recovery story; the old game wrote a
 *                                plain file, so we keep one too.
 *
 * The old build kept its sound toggles in SharedPreferences rather than in
 * save.json (that's the mobile/desktop split the `desktop_only.` key prefix
 * refers to), so they're read here and handed over alongside.
 */
object NativeState {

    private const val TAG = "NativeState"
    private const val LEGACY_SAVE = "save.json"
    private const val MIRROR = "sneakbit-save.json"

    /** The web build only ever sends a saveBlob envelope; anything near this
     *  size is a bug, not a save. */
    private const val MAX_MIRROR_BYTES = 256 * 1024

    /** SharedPreferences file and keys from the old build's AudioEngine. */
    private const val LEGACY_AUDIO_PREFS = "AudioSettings"
    private const val KEY_SFX_ENABLED = "kSoundEffectsEnabled"
    private const val KEY_MUSIC_ENABLED = "kMusicEnabled"

    /**
     * The envelope js/nativeBridge.js expects. Never throws: an unreadable file
     * is indistinguishable from an absent one as far as the game is concerned,
     * and the fallback still says "you're in a native shell", which is what
     * gates the mirror writes.
     */
    fun envelopeJson(context: Context): String {
        val envelope = JSONObject()
        envelope.put("platform", "android")
        envelope.put("mirror", readJsonObject(File(context.filesDir, MIRROR)) ?: JSONObject.NULL)

        // `legacy` only means something with a save in it — the audio prefs are
        // carried across as part of an import, never on their own.
        val legacySave = readJsonObject(File(context.filesDir, LEGACY_SAVE))
        if (legacySave != null) {
            val legacy = JSONObject()
            legacy.put("save", legacySave)
            legacy.put("audio", legacyAudio(context) ?: JSONObject.NULL)
            envelope.put("legacy", legacy)
        } else {
            envelope.put("legacy", JSONObject.NULL)
        }
        return envelope.toString()
    }

    private fun readJsonObject(file: File): JSONObject? = try {
        if (file.isFile && file.length() > 0) JSONObject(file.readText()) else null
    } catch (e: Exception) {
        Log.w(TAG, "unreadable ${file.name}: ${e.message}")
        null
    }

    /**
     * The old sound/music toggles, or null if the player never had them stored.
     * A key missing while the other is present defaults to enabled — that was
     * the old build's default.
     */
    private fun legacyAudio(context: Context): JSONObject? {
        val prefs = context.getSharedPreferences(LEGACY_AUDIO_PREFS, Context.MODE_PRIVATE)
        val hasSfx = prefs.contains(KEY_SFX_ENABLED)
        val hasMusic = prefs.contains(KEY_MUSIC_ENABLED)
        if (!hasSfx && !hasMusic) return null
        return JSONObject()
            .put("sfx", prefs.getBoolean(KEY_SFX_ENABLED, true))
            .put("music", prefs.getBoolean(KEY_MUSIC_ENABLED, true))
    }

    /**
     * Persist the web build's save mirror. Writes a temp file and renames, so a
     * kill mid-write can't leave a truncated save — the exact failure the old
     * Rust build's truncate-then-write was exposed to.
     */
    fun writeMirror(context: Context, text: String) {
        if (text.length > MAX_MIRROR_BYTES) return
        try {
            JSONObject(text)
        } catch (e: Exception) {
            Log.w(TAG, "refusing non-JSON mirror payload")
            return
        }
        val target = File(context.filesDir, MIRROR)
        val tmp = File(context.filesDir, "$MIRROR.tmp")
        try {
            tmp.writeText(text)
            if (!tmp.renameTo(target)) throw java.io.IOException("rename failed")
        } catch (e: Exception) {
            // A failed backup write must never disturb play; localStorage still
            // holds the save.
            Log.e(TAG, "mirror write failed", e)
            tmp.delete()
        }
    }
}

File: ios/SneakBit/NativeState.swift
//
//  NativeState.swift
//  SneakBit
//
//  The iOS half of the native save bridge (see js/nativeBridge.js) — the
//  counterpart of electron/nativeState.js and android/…/NativeState.kt.
//
//  Two files in the app container matter:
//
//    Documents/save.json           the save left by the Rust build this one
//                                  replaces. This ships as an update to the
//                                  same bundle id (it.curzel.bitscape), so the
//                                  container — and that file — survive the
//                                  update. The web build imports it once.
//
//    Documents/sneakbit-save.json  this build's own save, mirrored out of
//                                  localStorage. WebKit's storage is opaque and
//                                  has no recovery story; the old game wrote a
//                                  plain file, and players carry saves that
//                                  exist nowhere else, so we keep one too.
//
//  The old build also kept its sound toggles in UserDefaults rather than in
//  save.json (that's the mobile/desktop split the `desktop_only.` key prefix
//  refers to), so they're read here and handed over alongside.
//

import Foundation

enum NativeState {

    private static let legacySaveName = "save.json"
    private static let mirrorName = "sneakbit-save.json"

    /// The renderer only ever sends a saveBlob envelope. Anything near this
    /// size is a bug, not a save.
    private static let maxMirrorBytes = 256 * 1024

    /// UserDefaults keys from the old build's AudioEngine, verbatim.
    private static let kSoundEffectsEnabled = "kSoundEffectsEnabled"
    private static let kMusicEnabled = "kMusicEnabled"

    private static var documentsURL: URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }

    private static var legacySaveURL: URL { documentsURL.appendingPathComponent(legacySaveName) }
    private static var mirrorURL: URL { documentsURL.appendingPathComponent(mirrorName) }

    // MARK: - Reading

    /// The envelope js/nativeBridge.js expects, ready to serve. Never throws:
    /// an unreadable file is indistinguishable from an absent one as far as the
    /// game is concerned, and the fallback still says "you're in a native
    /// shell", which is what gates the mirror writes.
    static func envelopeData() -> Data {
        var envelope: [String: Any] = [
            "platform": "ios",
            "mirror": readJSONObject(at: mirrorURL) ?? NSNull(),
            "legacy": NSNull(),
        ]
        // `legacy` only means something with a save in it — the audio prefs are
        // carried across as part of an import, never on their own.
        if let save = readJSONObject(at: legacySaveURL) {
            let audio: Any = legacyAudio() ?? NSNull()
            envelope["legacy"] = ["save": save, "audio": audio]
        }
        guard let data = try? JSONSerialization.data(withJSONObject: envelope) else {
            return Data(#"{"platform":"ios","mirror":null,"legacy":null}"#.utf8)
        }
        return data
    }

    private static func readJSONObject(at url: URL) -> [String: Any]? {
        guard let data = try? Data(contentsOf: url),
              let parsed = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return parsed
    }

    /// The old sound/music toggles, or nil if the player never had them stored
    /// (a fresh install of the old build wrote both on first launch, so nil
    /// really does mean "never ran it"). A key that's missing while the other
    /// is present defaults to enabled — that was the old build's default.
    private static func legacyAudio() -> [String: Any]? {
        let defaults = UserDefaults.standard
        let sfx = defaults.object(forKey: kSoundEffectsEnabled) as? Bool
        let music = defaults.object(forKey: kMusicEnabled) as? Bool
        if sfx == nil && music == nil { return nil }
        return ["sfx": sfx ?? true, "music": music ?? true]
    }

    // MARK: - Writing

    /// Persist the renderer's save mirror. `.atomic` writes to a temp file and
    /// renames, so a kill mid-write can't leave a truncated save — the exact
    /// failure the old Rust build's truncate-then-write was exposed to.
    static func writeMirror(_ text: String) {
        let data = Data(text.utf8)
        guard data.count <= maxMirrorBytes,
              (try? JSONSerialization.jsonObject(with: data)) != nil
        else { return }
        do {
            try data.write(to: mirrorURL, options: .atomic)
        } catch {
            // A failed backup write must never disturb play; localStorage still
            // holds the save.
            NSLog("[NativeState] mirror write failed: \(error)")
        }
    }
}
